zipkintracing/tracing.go

```diff
@@ -0,0 +1,221 @@
+package zipkintracing
+
+import (
+	"fmt"
+	"github.com/labstack/echo/v4/middleware"
+	"net/http"
+	"strconv"
+
+	"github.com/labstack/echo/v4"
+	"github.com/openzipkin/zipkin-go"
+	zipkinhttp "github.com/openzipkin/zipkin-go/middleware/http"
+	"github.com/openzipkin/zipkin-go/model"
+	"github.com/openzipkin/zipkin-go/propagation/b3"
+	"github.com/openzipkin/zipkin-go/reporter"
+	zipkinHttpReporter "github.com/openzipkin/zipkin-go/reporter/http"
+)
+
+type (
+
+	// tracer struct
+	tracer struct {
+		ReportingURL string
+		ServiceName  string
+		tracer       *zipkin.Tracer
+		client       *zipkinhttp.Client
+		reporter     reporter.Reporter
+	}
+
+	// Tracer interface tracing
+	Tracer interface {
+		init() error
+		GetTraceServerFunc(tags map[string]string) echo.MiddlewareFunc
+		GetTraceServerFuncWithConfig(tags map[string]string, config TraceServerFuncConfig) echo.MiddlewareFunc
+		GetTraceProxyFunc(tags map[string]string) echo.MiddlewareFunc
+		GetTraceProxyFuncWithConfig(tags map[string]string, config TraceProxyFuncConfig) echo.MiddlewareFunc
+		TraceFunc(c echo.Context, spanName string, tags map[string]string) func()
+		StartChildSpan(c echo.Context, spanName string, tags map[string]string) (childSpan zipkin.Span)
+		Close() error
+		HTTPDoer
+	}
+
+	//HTTPDoer is a wrapper interface for http
+	HTTPDoer interface {
+		DoHTTP(c echo.Context, r *http.Request) (*http.Response, error)
+	}
+
+	//TraceServerFuncConfig config for TraceServerFunc
+	TraceServerFuncConfig struct {
+		Skipper middleware.Skipper
+	}
+
+	//TraceProxyFuncConfig config for TraceProxyFunc
+	TraceProxyFuncConfig struct {
+		Skipper middleware.Skipper
+	}
+)
+
+var (
+	//DefaultTraceServerFuncConfig default config for TraceServerFunc
+	DefaultTraceServerFuncConfig = TraceServerFuncConfig{
+		Skipper: middleware.DefaultSkipper,
+	}
+
+	//DefaultTraceProxyFuncConfig default config for TraceProxyFunc
+	DefaultTraceProxyFuncConfig = TraceProxyFuncConfig{
+		Skipper: middleware.DefaultSkipper,
+	}
+)
+
+// init function to initialize zipkin tracer
+func (t *tracer) init() error {
+	t.reporter = zipkinHttpReporter.NewReporter(t.ReportingURL)
+
+	// create our local service endpoint
+	endpoint, err := zipkin.NewEndpoint(t.ServiceName, "")
+	if err != nil {
+		return err
+	}
+	// initialize our tracer
+	tracer, err := zipkin.NewTracer(t.reporter, zipkin.WithLocalEndpoint(endpoint))
+	if err != nil {
+		return err
+	}
+	client, err := zipkinhttp.NewClient(tracer, zipkinhttp.ClientTrace(false))
+	if err != nil {
+		return err
+	}
+	t.client = client
+	t.tracer = tracer
+	return nil
+}
+
+// NewTracer returns instance of tracer with client initialized
+func NewTracer(zipkinURL, serviceName string) (Tracer, error) {
```

Comment:
Go has proverb `accept interfaces, return structs`   there is no need return interface. not like tracer has multiple implementations.   If developer needs interface he/she will create appropriate one in their code - where it is needed.